src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/VariantAnnotatorEngine.java

```diff
@@ -421,5 +465,189 @@ static AnnotationManager ofAllMinusExcluded(final List<String> annotationsToExcl
         }
 
     }
+    protected static class VAExpression {
+
+        public String fullName, fieldName;
+        public FeatureInput<VariantContext> binding;
+        public VCFInfoHeaderLine hInfo;
+
+        public VAExpression(String fullExpression, List<FeatureInput<VariantContext>> dataSourceList){
+            final int indexOfDot = fullExpression.lastIndexOf(".");
+            if ( indexOfDot == -1 ) {
+                throw new UserException.BadInput("The requested expression '"+fullExpression+"' is invalid, it should be in VCFFile.value format");
+            }
+
+            fullName = fullExpression;
+            fieldName = fullExpression.substring(indexOfDot+1);
+
+            final String bindingName = fullExpression.substring(0, indexOfDot);
+            for ( final FeatureInput<VariantContext> ds : dataSourceList ) {
+                if ( ds.getName().equals(bindingName) ) {
+                    binding = ds;
+                    break;
+                }
+            }
+        }
+    }
+
+    protected List<VAExpression> getRequestedExpressions() { return expressions; }
+
+    // select specific expressions to use
+    public void addExpressions(Set<String> expressionsToUse, List<FeatureInput<VariantContext>> dataSources, boolean expressionAlleleConcordance) {//, Set<VCFHeaderLines>) {
+        // set up the expressions
+        for ( final String expression : expressionsToUse ) {
+            expressions.add(new VAExpression(expression, dataSources));
+        }
+        this.expressionAlleleConcordance = expressionAlleleConcordance;
+    }
+
+    /**
+     * Handles logic for expressions for variant contexts. Used to add annotations from one vcf file into the fields
+     * of another if the variant contexts match sufficiently between the two files.
+     *
+     * @param vc  VariantContext to add annotations to
+     * @param features  FeatureContext object containing extra VCF features to add to vc
+     * @param ref  Reference context object corresponding to the region overlapping vc
+     * @param attributes  running list of attributes into which to place new annotations
+     */
+    private void annotateExpressions(final VariantContext vc,
+                                     final FeatureContext features,
+                                     final ReferenceContext ref,
+                                     final Map<String, Object> attributes){
+        Utils.nonNull(vc);
+
+        // each requested expression
+        for ( final VAExpression expression : expressions ) {
+            List<VariantContext> variantContexts = features.getValues(expression.binding, vc.getStart());
+
+            if (!variantContexts.isEmpty()) {
```

Comment:
There is too much nesting here, but you can reduce it with `if( variantContexts.isEmpty) { continue; }`